openhtf/__init__.py

```diff
@@ -151,15 +152,18 @@ def from_uid(cls, test_uid):
       raise UnrecognizedTestUidError('Test UID %s not recognized' % test_uid)
     return test
 
-  @property
-  def uid(self):
-    """Return a unique identifier for this Test instance.
+  def get_current_or_next_uid(self):
+    if self._executor is not None:
+      return self._executor.uid
+    return self._uid
+
+  def _increment_next_uid(self):
+    """Increments the next execution's UID.
 
-    Note that this identifier must be unique across Python invokations, so the
-    station_api UID is used as a prefix to guarantee that.  Test UID's need not
-    be unique across different hosts.
+    This identifier must be unique across invocations of execute() to
+    differentiate between different runs of the same test.
     """
-    return '%s:%s' % (station_api.STATION_API.UID, id(self))
+    self._uid = '%s:%s' % (station_api.STATION_API.UID, uuid.uuid4().hex)
```

Comment:
I found this logic with `get_current_or_next_uid` and `_increment_next_uid` hard to follow, and I think it's unnecessarily complex. We're saying that UIDs now belong to executions of a test and not the test itself, but it seems like the code is only halfway there.

It appears the only reason we need `self._uid` and the idea of a “next execution” is because of the call to `test.get_current_or_next_uid()` in `list_tests` of station_api.py. Could we instead decide that a test doesn't need to be included in the station API's list of tests if the test isn't executing yet? If so then `Test` could look something like this:

Remove `self._uid`, `get_current_or_next_uid`, and `_increment_next_uid`

```
@property
def uid(self):
  """Return UID of current execution, or None if the test is not executing."""
  if self._executor:
    return self._executor.uid

def make_uid(self):
  return '%s:%s' % (station_api.STATION_API.UID, uuid.uuid4().hex)

...

def execute(...):
  ...
  self._executor = core.TestExecutor(self._test_desc, self.make_uid(), ...)
```